synapse/lib/storm.py

```diff
@@ -2328,47 +2347,60 @@ def getArgParser(self):
         pars = Cmd.getArgParser(self)
         pars.add_argument('cond', help='The conditional value for the yield option.')
         pars.add_argument('genr', help='The generator function value that yields nodes.')
+        pars.add_argument('--size', default=None, help='The max number of times to iterate the generator.')
         return pars
 
     async def execStormCmd(self, runt, genr):
 
         if not isinstance(self.opts.genr, types.AsyncGeneratorType):
             raise s_exc.BadArg(mesg='The genr argument must yield nodes')
 
-        if self.runtsafe:
+        async def wrapgenr():
+            if self.runtsafe:
 
-            doyield = await s_stormtypes.tobool(self.opts.cond)
+                doyield = await s_stormtypes.tobool(self.opts.cond)
 
-            if doyield:
+                if doyield:
 
-                async for item in genr:
-                    await asyncio.sleep(0)
+                    async for item in genr:
+                        await asyncio.sleep(0)
 
-                async for item in self.opts.genr:
-                    yield item
-            else:
+                    async for item in self.opts.genr:
+                        yield item
+                else:
 
-                async for item in self.opts.genr:
-                    await asyncio.sleep(0)
+                    async for item in self.opts.genr:
+                        await asyncio.sleep(0)
 
-                async for item in genr:
-                    yield item
+                    async for item in genr:
+                        yield item
 
-            return
+                return
 
-        async for item in genr:
+            async for item in genr:
 
-            doyield = await s_stormtypes.tobool(self.opts.cond)
-            if doyield:
+                doyield = await s_stormtypes.tobool(self.opts.cond)
+                if doyield:
 
-                async for genritem in self.opts.genr:
-                    yield genritem
-            else:
+                    async for genritem in self.opts.genr:
+                        yield genritem
+                else:
 
-                async for genritem in self.opts.genr:
-                    await asyncio.sleep(0)
+                    async for genritem in self.opts.genr:
+                        await asyncio.sleep(0)
 
-                yield item
+                    yield item
+
+        size = await s_stormtypes.toint(self.opts.size, noneok=True)
```

Comment:
If the command isn't runtsafe, `self.opts.size` needs to be evaluated for each node. This also applies to the existing `self.opts.genr` check up top.